In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from scipy.stats import spearmanr
from scipy.stats import pearsonr

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/diamonds/diamonds.csv


The goal of this Python script is to find out if there is a relationship between the two variables the carat of a diamond and the price of a diamond, and to see what type of relationship exists between the two variables.

Hypothesis: I believe that there will be a positive relationship between the carat of a diamond and the price of a diamond, since in general we normally see the more carats a diamond has the higher of a price value the diamond tends to have as well.

To start I will perform a linear regression on the two variables carat of a diamond and price of a diamond. The goal of a regression is to investigate/predict if there is a relationship between the independent and dependent variable. A linear regression specifically is when data is modelled in a straight line in order to predict the value of a variable.

In [2]:
data = pd.read_csv("../input/diamonds/diamonds.csv")
data.head()

Unnamed: 0  carat      cut color clarity  depth  table  price     x     y  \
0           1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98   
1           2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84   
2           3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07   
3           4   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23   
4           5   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35   

      z  
0  2.43  
1  2.31  
2  2.31  
3  2.63  
4  2.75

To start we attribute arrays x and y to the columns we plan to analyze within this data set. We then check each column to confirm there are no values of 'nan' within the data set using two for loops. Finally we recieve our length of the data set which is 53,940 different values which meets the requirement of 10,000 values.

In [3]:
x = data['carat'].values
y = data['price'].values

new_x = []
new_y = []

for item in x:
    if str(item) != 'nan':
        new_x.append(item)

for item in y:
    if str(item) != 'nan':
        new_y.append(item)
        
print('The length of this dataset is:', len(new_x))

The length of this dataset is: 53940


Now in order to find r squared for our linear regression model, we must first start by finding out the equation for ypred. We will do this by first finding our new slope (b1), then finding our new intercept (b0).

In [4]:
mean_x = np.mean(new_x)
mean_y = np.mean(new_y)

num = 0
den = 0

m = len(new_x)

# Used to find variables for the new m for the ypred equation
for i in range(m):
    num += (new_x[i] - mean_x) * (new_y[i] - mean_y)
    den += (new_x[i] - mean_x) ** 2
    
b1 = num / den # solves for m in the ypred equation
b0 = mean_y - (b1 * mean_x) # solves for intercept in the ypred equation

print('ypred new slope:',b1)
print('ypred new intercept:',b0)

ypred new slope: 7756.425617967361
ypred new intercept: -2256.360580044546


This calculations we use to find r-squared. When find r-squared it is important that we also find the total sum of squares and the total sum of squares of residuals in order to make sure we have all the components necessary to perform the r-squared equation.

In [5]:
TSS = 0 #Total sum of squares
RSS = 0 #Total sum of square of residuals

for i in range(m):
    ypred = (b1 * new_x[i]) + b0 #y = mx + b which we will do for each value of i
    TSS += (new_y[i] - mean_y) ** 2
    RSS += (new_y[i] - ypred) ** 2

r2 = 1 - (RSS / TSS) #Formula for r-squared

print('r-squared value: ', r2)

r-squared value:  0.8493305264354793


Results of the Linear Regression:

What an r-squared value of .849 means is that 84.9% of the variation in the output variable can be explained by the input variable. In other words,the there is a strong positive relationship between the carat of a diamond and the price of a diamond, which can be attributed due to having an r-squared value of .849. This also means that our original hypothesis was indeed correct that there is a relationship between the carat of a diamond and the price of a diamond. This makes sense because as I may have mentioned earlier, the carat of a diamond does generally tend to be more expensive the higher the value of the carat is.

Next up I would like to perform a correlation on both the variables price of a diamond and the carat of a diamond. A correlation is the measure of how closely related two variables are to each other. Specifically we are going to be looking at both Person's and Spearman's correlation. Pearson's correlation is the strength of the linear relationship between two variables, so somewhat similar to regression. Spearman's correlation measures the strength and direction of monotonic (less-restrictive) association between two variables.

In [6]:
corr, _ = spearmanr(new_x, new_y)
corr, _ = pearsonr(new_x, new_y)

print('Pearsons Correlation:  ', corr)
print('Spearmans Correlation: ', corr)

Pearsons Correlation:   0.9215913011934773
Spearmans Correlation:  0.9215913011934773


Results of the Correlation Testing:

As we see from the results for both methods we get a correlation of .922. This is a very strong correlation, thus indicating that there is a strong positive relationship between the two variables carat of a diamond and the price of a diamond. Which means that as the carat of a diamond increases the price of the diamond increases as well. This result gives us more confirmation with the results of our previous linear regression test.